File: src/Algorithms/Lists/Sort_By_Score/sort-by-score.ts
const HIGHEST_POSSIBLE_SCORE = 100;

// O(n) time
// O(n) space
const solution = (unsortedScores: number[]) => {
  const sorted = new Array(HIGHEST_POSSIBLE_SCORE).fill(0);

  for (let i = 0; i < unsortedScores.length; i++) {
    sorted[unsortedScores[i]] += 1;
  }

  return sorted.reduceRight((acc, item, index) => {
    if (item > 0) {
      let i = 0;
      while (i < item) {
        acc.push(index);
        i++;
      }
    }

    return acc;
  }, []);
};

export const solutions = {
  Solution: solution,
};

File: src/Algorithms/Lists/Sort_By_Score/sort_by_score.py


# O(n) time
# O(1) space
def solution(list):
    highest_score = 100
    sorted_scores = []
    possible_scores = [0] * highest_score

    for score in list:
        possible_scores[score] += 1

    for amount in possible_scores[::-1]:
        highest_score -= 1
        while amount > 0:
            sorted_scores.append(highest_score)
            amount -= 1

    return sorted_scores
